pkg/network/master/egressip.go

```diff
@@ -0,0 +1,95 @@
+package master
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	utilwait "k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/util/retry"
+
+	"github.com/openshift/origin/pkg/network/common"
+	networkinformers "github.com/openshift/origin/pkg/network/generated/informers/internalversion/network/internalversion"
+	networkclient "github.com/openshift/origin/pkg/network/generated/internalclientset"
+)
+
+type egressIPManager struct {
+	sync.Mutex
+
+	tracker            *common.EgressIPTracker
+	networkClient      networkclient.Interface
+	hostSubnetInformer networkinformers.HostSubnetInformer
+
+	updatePending bool
+	updatedAgain  bool
+}
+
+func newEgressIPManager() *egressIPManager {
+	eim := &egressIPManager{}
+	eim.tracker = common.NewEgressIPTracker(eim)
+	return eim
+}
+
+func (eim *egressIPManager) Start(networkClient networkclient.Interface, hostSubnetInformer networkinformers.HostSubnetInformer, netNamespaceInformer networkinformers.NetNamespaceInformer) {
+	eim.networkClient = networkClient
+	eim.hostSubnetInformer = hostSubnetInformer
+	eim.tracker.Start(hostSubnetInformer, netNamespaceInformer)
+}
+
+func (eim *egressIPManager) UpdateEgressCIDRs() {
+	eim.Lock()
+	defer eim.Unlock()
+
+	if eim.updatePending {
+		eim.updatedAgain = true
+	} else {
+		eim.updatePending = true
+		go utilwait.PollInfinite(time.Second, eim.maybeDoUpdateEgressCIDRs)
```

Comment:
Will you have a deadlock here? eim.Lock is already held, if I understand this correctly

Reply:
between UpdateEgressCIDRs() and maybeDoUpdateEgressCIDRs()? No, because the latter is called in a goroutine, after the former returns

Reply:
@danwinship so basically, if UpdateEgressCIDRs() gets called multiple times in the same second, the next maybeDoUpdateEgressCIDRs() will be skipped and it'll wait until it gets at least one second of no activity before it actually does the stuff in maybeDoUpdateEgressCIDRs(), right?

Reply:
ah yes, makes sense. assuming I understand this correctly, nothing should be skipped because it will just wait for the lock.